### PR TITLE
Visual Search API: add image visual search SDK and tests

### PR DESCRIPTION
**Pull request by shakao**
## Description

Adding Image Visual Search SDK generated from Swagger spec ([pull request](https://github.com/Azure/azure-rest-api-specs/pull/2872)) and tests. 

The API takes a POST request with an image binary in the body (using multipart/form-data), and has a restriction that the filename of the image must be exactly the string "image". The SDK was modified to enforce this requirement in [/BingVisualSearch/Generated/Images.cs](https://github.com/shakao/azure-sdk-for-net/blob/38630f7f7f6e9b52f2e9a8e6cb3a770356ba605a/src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Generated/Images.cs#L324).

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/Microsoft.Azure.CognitiveServices.Search.VisualSearch.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+<Import Project="$([MSBuild]::GetPathOfFileAbove('AzSdk.test.reference.props'))" />
+  <PropertyGroup>
+    <Description>Microsoft.Azure.CognitiveServices.Search.VisualSearch.Tests Class Library</Description>
+    <AssemblyName>Microsoft.Azure.CognitiveServices.Search.VisualSearch.Tests</AssemblyName>
+    <VersionPrefix>1.1.0-preview</VersionPrefix>
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/Microsoft.Azure.CognitiveServices.Search.VisualSearch.Tests.csproj]: Please set package version for test projects to be `1.0.0`

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/Microsoft.Azure.CognitiveServices.Search.VisualSearch.Tests.csproj [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Management.ResourceManager" Version="1.6.0-preview" />
+    <ProjectReference Include="..\BingVisualSearch\Microsoft.Azure.CognitiveServices.Search.VisualSearch.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="SessionRecords\**\*.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+	<None Update="TestImages\*.jpg">
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/Microsoft.Azure.CognitiveServices.Search.VisualSearch.Tests.csproj]: nit: Fix indentation
Also check if these files are a part of your build

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/TestModels/KnowledgeRequest.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+namespace Microsoft.Azure.CognitiveServices.Search.VisualSearch.TestModels
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/TestModels/KnowledgeRequest.cs]: Please add the license headers

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/generate.cmd [hunk 4]
```diff
@@ -0,0 +1,7 @@
+::
+:: Microsoft Azure SDK for Net - Generate library code
+:: Copyright (C) Microsoft Corporation. All Rights Reserved.
+::
+
+@echo off
+call %~dp0..\..\..\..\..\..\tools\generate.cmd cognitiveservices/data-plane/VisualSearch %*
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/generate.cmd]: Please create a generate.ps1 file similar to [this](https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/src/SDKs/CognitiveServices/dataPlane/Language/Language/Language/generate.ps1) and regenerate the code

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Reflection;
+using System.Resources;
+
+[assembly: AssemblyTitle("Microsoft Azure Cognitive Services Visual Search Client Library")]
+[assembly: AssemblyDescription("Provides API functions for consuming the Microsoft Azure Cognitive Services Visual Search API.")]
+
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.1.0.0")]
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Properties/AssemblyInfo.cs]: Any reason you wish to start versioning from `1.1.0.0` here

**Review by dsgouda — CHANGES_REQUESTED** [review]
Please generate the code using generate.cmd or generate.ps1 and commit all generated files
Please run `msbuild build.proj /t:build /p:Scope=SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch`

and commit all the changes

**Comment by shakao** [comment]
@dsgouda hey! so i regenerated the code with the powershell script and ran msbuild again, but there weren't any changes from the initially committed files? is there something that you were expecting to be updated? i've made the other requested changes

**Comment by dsgouda** [comment]
@shakao Yes, there should be a .txt file generated which logs information related to the code generation
Do you see that file being generated when you run generate.cmd?

**Comment by shakao** [comment]
@dsgouda oh! yes! sorry, i didn't commit it because i thought _metadata was a local thing but i'll add it in now

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Microsoft.Azure.CognitiveServices.Search.VisualSearch.csproj [hunk 6]
```diff
@@ -0,0 +1,21 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <!-- Please do not move/edit code below this line -->
+  <Import Project="$([MSBuild]::GetPathOfFileAbove('AzSdk.reference.props'))" />
+  <!-- Please do not move/edit code above this line -->
+
+  <PropertyGroup>
+    <PackageId>Microsoft.Azure.CognitiveServices.Search.VisualSearch</PackageId>
+    <Description>This client library provides access to the Microsoft Cognitive Services Image Visual Search API.</Description>
+    <VersionPrefix>1.1.0-preview</VersionPrefix>
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Microsoft.Azure.CognitiveServices.Search.VisualSearch.csproj]: We strongly recommend starting versioning from `0.9.0-preview` for preview packages

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Generated/SdkInfo_VisualSearchAPI.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Generated/SdkInfo_VisualSearchAPI.cs]: Please run `msbuild build.proj /t:build /p:Scope=SDKs\/CognitiveServices/dataPlane/Search/BingVisualSearch/` and commit the `.props` file generated

**Review by dsgouda — CHANGES_REQUESTED** [review]
Left a couple of comments, looks good for the most part

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Generated/SdkInfo_VisualSearchAPI.cs (hunk 7 above)
Comment (shakao) [reply src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Generated/SdkInfo_VisualSearchAPI.cs]: hmm, i see a .props in the /obj output on building--should i move that file somewhere and commit it?

Comment (dsgouda) [reply src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch/Generated/SdkInfo_VisualSearchAPI.cs]: @shakao 
Talked to @shahabhijeet I think we can skip the .props file for now and we will investigate later why the script might have failed (the `.props` file is more important for the management sdks)

**Review by dsgouda — APPROVED** [review]

**Review by dsgouda — CHANGES_REQUESTED** [review]
Almost forgot, please fix the version numbers

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/VisualSearchTests.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            {
+                HttpMockServer.Initialize(this.GetType().FullName, "VisualSearchWithBinary");
+
+                IVisualSearchAPI client = new VisualSearchAPI(new ApiKeyServiceClientCredentials(SubscriptionKey), HttpMockServer.CreateInstance());
+
+                using (FileStream stream = new FileStream(Path.Combine("TestImages", "image.jpg"), FileMode.Open))
+                {
+                    KnowledgeRequest KnowledgeRequest = new KnowledgeRequest();
+
+                    var resp = client.Images.VisualSearchMethodAsync(image: stream, knowledgeRequest: JsonConvert.SerializeObject(KnowledgeRequest)).Result;
```
Comment (shahabhijeet) [inline src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/VisualSearchTests.cs]: @shakao why are you not using the new function from your customization layer.
Why KnowledgeRequest is still being serialized?

**Review by shahabhijeet — CHANGES_REQUESTED** [review]
I don't see where you are using the new function that you introduced in the customization layer in your tests.
Also please point me where have you created two customization functions?

File: src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/VisualSearchTests.cs (hunk 8 above)
Comment (shakao) [reply src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/VisualSearchTests.cs]: the new test is "VisualSearchWithKnowledgeRequestObject" (line 102)! these tests are there for the serialized functions

Comment (shakao) [reply src/SDKs/CognitiveServices/dataPlane/Search/BingVisualSearch/BingVisualSearch.Tests/VisualSearchTests.cs]: and the extensions are in the /Customizations folder (IImages.cs, Images.cs, ImageExtensions.cs)

**Review by shahabhijeet — APPROVED** [review]
Once CI passes, this is good to be merged

**Review by dsgouda — APPROVED** [review]

**Pull request merged**